src/storage/gcs/storage.c

```diff
@@ -0,0 +1,949 @@
+
+            cryptoError(EVP_DigestSignFinal((EVP_MD_CTX *)sign, bufPtr(signature), &signatureLen) <= 0, "unable to finalize");
+
+            // Add signature
+            strCatChr(result, '.');
+            strCatEncode(result, encodeBase64Url, signature);
+        }
+        FINALLY()
+        {
+            BIO_free((BIO *)bio);
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+            EVP_MD_CTX_cleanup((EVP_MD_CTX *)sign);
+#else
+            EVP_MD_CTX_free((EVP_MD_CTX *)sign);
+#endif
+            EVP_PKEY_free((EVP_PKEY *)privateKey);
+        }
+        TRY_END();
+    }
+    MEM_CONTEXT_TEMP_END();
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+typedef struct
+{
+    String *tokenType;
+    String *token;
+    time_t timeExpire;
+} StorageGcsAuthTokenResult;
+
+static StorageGcsAuthTokenResult
+storageGcsAuthToken(StorageGcs *this, time_t timeBegin)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(STORAGE_GCS, this);
+        FUNCTION_TEST_PARAM(TIME, timeBegin);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+    ASSERT(timeBegin > 0);
+
+    StorageGcsAuthTokenResult result = {0};
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        String *content = strNewFmt(
+            "grant_type=urn%%3Aietf%%3Aparams%%3Aoauth%%3Agrant-type%%3Ajwt-bearer&assertion=%s",
+            strZ(storageGcsAuthJwt(this, timeBegin)));
+
+        HttpHeader *header = httpHeaderNew(NULL);
+        httpHeaderAdd(header, HTTP_HEADER_HOST_STR, httpUrlHost(this->authUrl));
+        httpHeaderAdd(header, HTTP_HEADER_CONTENT_TYPE_STR, HTTP_HEADER_CONTENT_TYPE_APP_FORM_URL_STR);
+        httpHeaderAdd(header, HTTP_HEADER_CONTENT_LENGTH_STR, strNewFmt("%zu", strSize(content)));
+
+        HttpRequest *request = httpRequestNewP(
+            this->authClient, HTTP_VERB_POST_STR, httpUrlPath(this->authUrl), NULL, .header = header, .content = BUFSTR(content));
+        HttpResponse *response = httpRequestResponse(request, true);
+
+        KeyValue *kvResponse = jsonToKv(strNewBuf(httpResponseContent(response)));
+
+        // Check for an error
+        const String *error = varStr(kvGet(kvResponse, GCS_JSON_ERROR_VAR));
+
+        if (error != NULL)
+        {
+            const String *description = varStr(kvGet(kvResponse, GCS_JSON_ERROR_DESCRIPTION_VAR));
+            CHECK(description != NULL);
+
+            THROW_FMT(ProtocolError, "unable to get authentication token: [%s] %s", strZ(error), strZNull(description));
+        }
+
+        MEM_CONTEXT_PRIOR_BEGIN()
+        {
+            // Get token
+            result.tokenType = strDup(varStr(kvGet(kvResponse, GCS_JSON_TOKEN_TYPE_VAR)));
+            CHECK(result.tokenType != NULL);
+            result.token = strDup(varStr(kvGet(kvResponse, GCS_JSON_ACCESS_TOKEN_VAR)));
+            CHECK(result.token != NULL);
+
+            // Get expiration
+            const Variant *const expiresIn = kvGet(kvResponse, GCS_JSON_EXPIRES_IN_VAR);
+            CHECK(expiresIn != NULL);
```

Comment:
If we're going to error anyway, would it be more helpful to remove the CHECKs so we can see all the values are set or not and combine them into an IF statement:
```
if (result.tokenType == NULL || result.token == NULL || expiresIn == NULL) 
    THROW_FMT(ProtocolError, "tokenType %s, token %s, expiresIn %s", strZNull(result.tokenType), strZNull(result.token), strZNull(expiresIn));
```

Reply:
I originally did that but it generates a lot of tests and didn't seem worth it. If Google starts returning bad data we'll just need to debug that.

Also, the user is unlikely to want to send us these values since they might allows access to their bucket. And we avoid putting secrets in log messages.